### PR TITLE
Unsaferow serde with memory management

### PR DESCRIPTION
**Pull request by beroyfb**
-Needed for batch based shuffle services.
-Also includes a full e2e round trip fuzzer test.

### CODE REVIEW

File: velox/serializers/UnsafeRowSerde.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    remainingSize = block.size() - (buffer - originalBuffer);
+    if (remainingSize < 0) {
+      return false;
+    }
+  }
+  return true;
+}
+} // namespace
+
+std::optional<size_t> UnsafeRowVectorSerde::serializeKeys(
```
Comment (miaoever) [inline velox/serializers/UnsafeRowSerde.cpp:64]: The keys could be serialized by different serializer (with different format not necessary the unsaferow format) - i.e, Cosco is using it's own key serializer. It'd be better to make the key serializer pluggable.

Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp]: Very good idea! I moved it to a factory!

Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp]: @miaoever lmk if you have more comments on this PR

File: velox/serializers/UnsafeRowSerde.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      vector->type(), vector.get(), row);
+
+  // Check if the current allocated buffer is large enough for this row
+  if (rowBuffer_.data() == nullptr || rowSize > rowBuffer_.size()) {
+    // Need more memory if we are serializing keys as well
+    // We allocate at least twice the needed size so if the new rows are
+    // bigger we do not have re-allocate
+    // In case there are keys we allocate three times the size
+    // The previously allocated buffer will be deallocated automatically
+    auto allocationSize = keys_.has_value() ? rowSize * 3 : rowSize * 2;
```
Comment (kgpai) [inline velox/serializers/UnsafeRowSerde.cpp:78]: sizeof 1 row seems a poor metric no, shouldnt it be probable size of all rows ?

File: velox/vector/VectorBatch.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+class VectorSerde {
+ public:
+  VectorSerde(const std::unique_ptr<VectorKeySerializer>& keySerializer)
+      : keySerializer_{std::move(keySerializer)} {}
+  virtual ~VectorSerde() = default;
+
+  /// Must be called every time a new vector will be processed
+  /// \param vector the new vector to be serialized
+  virtual void initialize(const std::shared_ptr<RowVector>& vector) = 0;
```
Comment (kgpai) [inline velox/vector/VectorBatch.h:67]: Cant this be moved to constructor and some static factory cretaed for the Serde so people do not have to remember this kind of state ?

File: velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  UnsafeRowBatchSerdeFuzzTests() {}
+
+  void resetBuffer() {
+    blockUsedSize_ = 0;
+  }
+
+  void addRowToBuffer(
+      const std::string_view& serializedKeys,
+      const std::string_view& serializedRow) {
+    if (blockUsedSize_ + serializedKeys.size() + serializedRow.size() +
```
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp:46]: What that means is if the current buffer size plus the new key and value plus the two extra size_t (needed to write the sizes of key and value) is bigger than the block size. Just checking if the block has capacity for the new key/value.

File: velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      std::memcpy(newBuffer, buffer_, blockUsedSize_);
+      // Switch the old block with the new one
+      bufferPtr_ = newBufferPtr;
+      buffer_ = newBuffer;
+      blockSize_ = newBlockSize;
+    }
+    // Add the row (and keys) to the block
+    auto buffer = buffer_ + blockUsedSize_;
+    *((size_t*)buffer) = serializedKeys.size();
+    buffer += sizeof(size_t);
```
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp:64]: We write the size of the keys as well in the block.

File: velox/serializers/UnsafeRowSerde.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  char* buffer = const_cast<char*>(rowBuffer.data());
+  char* originalBuffer = buffer;
+  size_t keySize = 0;
+  buffer += sizeof(keySize);
+
+  auto fieldTypes =
+      std::dynamic_pointer_cast<const RowType>(vector->type())->children();
+  // Traverse the keys and serialize them
+  for (auto index : keys) {
+    if (index < 0 || index > fieldTypes.size()) {
```
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:37]: Not sure what you mean here. I'm passing std::vector& keys

File: velox/serializers/UnsafeRowSerde.cpp (hunk 2 above)
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:78]: This function is only for writing one row to the output (like shuffle).

File: velox/vector/VectorBatch.h (hunk 3 above)
Comment (beroyfb) [reply velox/vector/VectorBatch.h:67]: No, the serializer can go through multiple vectors.

File: velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/row/tests/UnsafeRowBatchSerdeFuzzTests.cpp:62]: Added enough comments to clarify that.

File: velox/vector/VectorBatch.h (hunk 3 above)
Comment (beroyfb) [reply velox/vector/VectorBatch.h:67]: You remember this takes care of pre-loading of the input vector and after that we can use the size calculator safely. Yes, the user is supposed to call initialize before serializing any vector. The pattern in which it works is like this:
```<constructor> <initialize> <serialize row x> ... <serialize row y> <initialize> <serialize row xx> ... <serialize row yy> ...```
I can try to change it so each time we need to ```initialize``` we create a new object but all the methods there are static.

File: velox/serializers/UnsafeRowSerde.cpp (hunk 2 above)
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:78]: What do you suggest? The buffer is supposed to hold one row at a time though. I can change the numbers to 10 and 5 if that's better.

File: velox/serializers/UnsafeRowSerde.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  if (rowBuffer_.data() == nullptr || rowSize > rowBuffer_.size()) {
+    // Need more memory if we are serializing keys as well
+    // We allocate at least twice the needed size so if the new rows are
+    // bigger we do not have re-allocate
+    // In case there are keys we allocate three times the size
+    // The previously allocated buffer will be deallocated automatically
+    auto allocationSize = keys_.has_value() ? rowSize * 3 : rowSize * 2;
+    rowBufferPtr_ = AlignedBuffer::allocate<char>(allocationSize, pool_, true);
+    if (rowBufferPtr_ == nullptr) {
+      return BatchSerdeStatus::AllocationFailed;
```
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:81]: For AllocationFailed most likely it has to throw an exception.

File: velox/serializers/UnsafeRowSerde.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    std::memset(const_cast<char*>(rowBuffer_.data()), 0, allocationSize);
+  }
+
+  size_t rowOffset = 0;
+  // first, writing the keys if any
+  if (keys_.has_value() && keySerializer_ != nullptr) {
+    auto offset = keySerializer_->serializeKeys(
+        vector, keys_.value(), row, rowBuffer_, serializedKeys);
+    if (!offset.has_value()) {
+      return BatchSerdeStatus::KeysSerializationFailed;
```
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:94]: That's a good idea actually.

File: velox/vector/VectorBatch.h (hunk 3 above)
Comment (beroyfb) [reply velox/vector/VectorBatch.h:67]: I will call it ```reset```

File: velox/serializers/UnsafeRowSerde.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kgpai) [reply velox/serializers/UnsafeRowSerde.cpp:82]: So why not throw an exception here.. are there cases where we dont want that ?

File: velox/vector/VectorBatch.h [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  DeserializationError,
+  KeysSerializationFailed,
+  KeyIndexOutOfRange,
+  SizesDeserializationError
+};
+
+class VectorKeySerializer {
+ public:
+  /// Deserializes a set of keys in a vector an aggregated key
+  /// \param vector the input row vector
```
Comment (kgpai) [inline velox/vector/VectorBatch.h:45]: For consistency's sake can we get rid of the \param's ; I have only seen it in one other place apart from UnsafeRowSerializer and while not explicitly mentioned I dont think it meets our coding guidelines.. (Keep the helpful descriptions on the parameters).

**Review by kgpai — APPROVED** [review]

File: velox/serializers/UnsafeRowSerde.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (beroyfb) [reply velox/serializers/UnsafeRowSerde.cpp:82]: Makes sense

**Pull request closed without merging**